content/blog/2019-10-19-loop-invariant-code-motion/index.md

```diff
@@ -0,0 +1,297 @@
+
+# Motion
+
+### Hoisting instructions
+
+Once we have determined which instructions are loop invariant, we check if we can actually hoist those instructions out of the loop. The following conditions must be met:
+
+* `condition 1:` instruction dominates all loop exits
+
+* `condition 2:` instruction is the only definition for that variable in the loop
+
+* `condition 3:` definition dominates all of its uses
+
+After identifying all three criteria, an instruction is labeled in a Boolean map, as `true` if hoistable, and `false` if not. 
+Instructions are removed from their native basic blocks if marked as `true`. 
+
+### Preheader generation and target renaming
+
+We gather hoisted instructions into a new basic preheader block executed before the loop header.
+This means we have to rename jumps that target the loop header to target the preheader instead.
+We only rename jumps that target the loop header if they are outside the loop. This ensures that the preheader is only to be executed when we first enter the loop.
+
+
+### Instruction list generation
+
+
+Next is to generate a new list of instructions for the function body. To do this, we initially order the basic blocks. The order must follow several constraints: the first and last basic blocks of the original program remain in place. Fall-throughs between basic blocks and the original function are also respected.
+
+Here, we ran into a few bugs. 🐝🐛🦋🐌🐞🐜🦗🕷🦂
+
+The existing code in the repository that computes dominator trees throws an exception when a basic block is unreachable. To solve this, we removed the unreachable block from all data structures. This implements a very limited form of dead code eliminations as a bonus! ☠️
+
+When we have an order for basic blocks we concatenate them to create a new list of instructions. Further post-processing to remove empty preheaders is then executed. Finally we have our new instruction list!
+
+# Evaluation
+We evaluated our optimization on a suite of benchmarks by instrumenting the `bril` interpreter to count the number of instructions it executes. Our optimization are evaluated this way, as opposed to comparing execution times. This avoids measurement biases that might arise through timing. 
+
+By abstracting away performance in terms of the number of instructions executed, we give a fair comparison between non-optimized and optimized `bril` programs: no matter the environment in which they are executed, a `bril` program that executes fewer instructions performs better than an equivalent `bril` program that executes more instructions.
```

Comment:
I think you can just say Bril instead of `bril`—we've been using it as a proper English name rather than considering it "code."